### ISSUE TITLE
Wheel scroll problem on nested JTabbedPane for tabPlacement other than TOP

### ISSUE THREAD

**Issue by ron190** [aterai/java-swing-tips#3]
I'm encountering a strange Swing problem on basic nested **JTabbedPane**, and I wonder if you know what is happening.

Wheel scrolling is added to tabbedPanes and all is fine when tabs are on top of panes. But **behavior doesn't work has expected when tab placement is on bottom or on left/right**, it's the parent tabbedPane that is scrolled instead of the wheeled tabbedPane.

Have you encountered this problem already and have you thought about a fix?

Many thanks.

> I made some debugging, it appears the **e.getSource()** of MouseWheelListener does not return the wheeled tabbedPane which is on SwingConstants.BOTTOM/LEFT/RIGHT, instead it returns the parent tabbedPane.
> 
> Here is the animated Gif of the problem:
> - the two tabs on top are fine with tabPlacement on default (SwingConstants.TOP)
> - the tabs on the left and on the bottom are scrolling their parents.
> 

> **Gif [Imgur](https://i.imgur.com/SIQIN9r.gifv): parent tabbedPane is changing when nested left/bottom tabbedPane is wheel-scrolled**
> 
> ![image](https://user-images.githubusercontent.com/1971480/79260183-2934bf80-7e8e-11ea-9bd4-227212a89577.png)
> 
> 
> 
> 
> **Code**
> 
> ```java
> /**
>  * Mousewheel allows to navigate to next/previous tab.
>  */
> public class TabbedPaneMouseWheelScroller implements MouseWheelListener {
>     
>     /**
>      * Log4j logger sent to view.
>      */
>     private static final Logger LOGGER = Logger.getRootLogger();
>     
>     @Override
>     public void mouseWheelMoved(MouseWheelEvent e) {
>         
>         JTabbedPane tabPane = (JTabbedPane) e.getSource();
> 
>         int dir = -e.getWheelRotation();
>         int selIndex = tabPane.getSelectedIndex();
>         int maxIndex = tabPane.getTabCount() - 1;
>         
>         if ((selIndex == 0 && dir < 0) || (selIndex == maxIndex && dir > 0)) {
>             
>             selIndex = maxIndex - selIndex;
>             
>         } else {
>             
>             selIndex += dir;
>         }
>         
>         if (0 <= selIndex && selIndex < tabPane.getTabCount()) {
>             
>             // Fix #54575: NullPointerException on setSelectedIndex()
>             try {
>                 tabPane.setSelectedIndex(selIndex);
>             } catch (NullPointerException err) {
>                 LOGGER.error(err, err);
>             }
>         }
>     }
> }
> ```

**Comment by aterai** [comment]
Hi ron109,
I misread your email and mistook it for a scrolling bug in the [DnDExportTabbedPane](https://github.com/aterai/java-swing-tips/tree/master/DnDExportTabbedPane)  and instructed this github to send a bug report.
I'm terribly sorry.

But this is an interesting issue, so let me continue here.

----
I am writing the following reproduction code, but it is not well made.
```
import java.awt.*;
import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;
import java.util.stream.IntStream;
import javax.swing.*;

public class TabbedPaneMouseWheelTest {
  private Component makeUI() {
    TabbedPaneMouseWheelScroller scroller = new TabbedPaneMouseWheelScroller();

    JTabbedPane jtp1 = new JTabbedPane(SwingConstants.TOP, JTabbedPane.SCROLL_TAB_LAYOUT);
    jtp1.addMouseWheelListener(scroller);
    JTabbedPane jtp2 = new JTabbedPane(SwingConstants.BOTTOM, JTabbedPane.SCROLL_TAB_LAYOUT);
    jtp2.addMouseWheelListener(scroller);
    JTabbedPane jtp3 = new JTabbedPane(SwingConstants.TOP, JTabbedPane.SCROLL_TAB_LAYOUT);
    jtp3.addMouseWheelListener(scroller);
    JTabbedPane jtp4 = new JTabbedPane(SwingConstants.LEFT, JTabbedPane.SCROLL_TAB_LAYOUT);
    jtp4.addMouseWheelListener(scroller);

    jtp1.addTab("JTabbedPane TOP", jtp2);
    jtp1.addTab("JTabbedPane 00001", new JLabel("JLabel1"));

    jtp2.addTab("JTabbedPane BOTTOM", jtp3);
    jtp2.addTab("JTabbedPane 00002", new JLabel("JLabel2"));
    jtp2.addTab("JTabbedPane 00003", new JLabel("JLabel3"));

    jtp3.addTab("JTabbedPane LEFT", jtp4);
    jtp3.addTab("JTabbedPane 00004", new JLabel("JLabel4"));
    jtp3.addTab("JTabbedPane 00005", new JLabel("JLabel5"));

    IntStream.range(0, 20).forEach(i ->
        jtp4.addTab("title" + i, new JScrollPane(new JTextArea("label" + i))));

    return jtp1;
  }

  public static void main(String[] args) {
    EventQueue.invokeLater(() -> {
      JFrame frame = new JFrame();
      frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
      frame.getContentPane().add(new TabbedPaneMouseWheelTest().makeUI());
      frame.setSize(320, 240);
      frame.setLocationRelativeTo(null);
      frame.setVisible(true);
    });
  }
}

class TabbedPaneMouseWheelScroller implements MouseWheelListener {
  @Override
  public void mouseWheelMoved(MouseWheelEvent e) {

    JTabbedPane tabPane = (JTabbedPane) e.getSource();

    int dir = -e.getWheelRotation();
    int selIndex = tabPane.getSelectedIndex();
    int maxIndex = tabPane.getTabCount() - 1;

    if (selIndex == 0 && dir < 0 || selIndex == maxIndex && dir > 0) {
      selIndex = maxIndex - selIndex;
    } else {
      selIndex += dir;
    }

    if (0 <= selIndex && selIndex < tabPane.getTabCount()) {
      try {
        tabPane.setSelectedIndex(selIndex);
      } catch (NullPointerException err) {
        // LOGGER.error(err, err);
      }
    }
  }
}
```

**Comment by ron190** [comment]
No problem, in fact I searched this use case on your website before sending it to you, because you manage so many Swing tips already that I thought you may be interested in this one.

**Comment by ron190** [comment]
My turn to be sorry @aterai, you posted this code snippet and I thought it was just to reproduce the bug. Because I haven't noticed any difference with the code I use, so I didn't tested it.

Until today... and I discovered your code works. 

The only difference with my code is that you use a single instance of `new TabbedPaneMouseWheelScroller()` with `addMouseWheelListener()`.

Whereas I use a new `WheelListener` in the constructor of each tabbedPane:
```java
public class TabbedPaneWheeled extends JTabbedPane {
    public TabbedPaneWheeled() {
        this.addMouseWheelListener(new TabbedPaneMouseWheelScroller());
```
The problem I encountered on parent tabs was maybe that a static `AWTEventMulticaster` registers the `WheelListener`, but it's not really clear to me:
```java
public synchronized void addMouseWheelListener(MouseWheelListener l) {
      mouseWheelListener = AWTEventMulticaster.add(mouseWheelListener,l);
}
```

Anyway I changed my code to have a single WheelListener and now it works!
Many thanks for your time ;)
